docs/src/modules/ROOT/pages/constraints-and-score/constraints-and-score.adoc

```diff
@@ -1511,6 +1511,47 @@ the tuple `(SomePerson, USER)` is sent downstream twice.
 See <<constraintStreamsDealingWithDuplicateTuplesUsingDistinct,`distinct()`>> for how to deal with duplicate tuples.
 ====
 
+[[constraintStreamsConcat]]
+==== Concat
+
+Concat allows you to create a constraint stream containing tuples of two constraint streams of the same <<constraintStreamsCardinality,cardinality>>.
+If <<constraintStreamsJoin,join>> acts like a cartesian product of two lists, concat acts like a concatenation of two lists.
+Unlike union of sets, concatenation of lists repeat duplicated elements.
+If the two constraint stream parents share tuples, they will be repeated downstream.
+If this is undesired, use <<constraintStreamsDealingWithDuplicateTuplesUsingDistinct,distinct>>.
+
+image::constraints-and-score/constraintStreamConcat.png[align="center"]
+
+For example, to count the total number of cats and dogs:
+
+[source,java,options="nowrap"]
+----
+    private Constraint countCatsAndDogs(ConstraintFactory constraintFactory) {
+        return constraintFactory.forEach(Employee.class)
+         .filter(employee -> employee.hasDog())
+         .concat(constraintFactory.forEach(Employee.class)
+                     .filter(employee -> employee.hasCat()))
```

Comment:
I am not sure this is a good example. The first question once a smart person sees this is: why not just use a much simple `.filter(hasCat || hasDog)`?

I think we need an example which doesn't have an obvious simpler faster solution.

Reply:
The obvious example would be to implement group by on an outer join, like this:

```java
factory.forEach(A.class)
    .join(B.class, ...)
    .concat(factory.forEach(A.class)
                               .ifNotExists(B.class, ...)
                               .expand(a -> null))
    .groupBy((a,b) -> a, conditionally((a,b) -> b != null, countBi())
    ...
```

Although the example would lead to the question "why no outer join" since that would simplify the code too

```java
factory.forEach(A.class)
    .outerJoin(B.class, ...)
    .groupBy((a,b) -> a, conditionally((a,b) -> b != null, countBi())
    ...
```

The other thing it can be used for is or'ing Joiners, which is unrelated to outer joins

```java
factory.forEach(A.class)
    .join(B.class, equal(a -> a.value, b -> b.value))
    .concat(factory.forEach(A.class)
                               .join(B.class, equal(a -> a.group, b -> b.group)))
    ...
```

Reply:
If we're doing concat for the minimal constraint, why not show that? Maybe there's a simplified variant that isn't too much code while still showing the point?

I'm not worried if people start asking about outer join after they see this. First of all, we have a decent answer, this is faster because it doesn't need to do any sort of indexing. Second, if they do, and if there's enough of them, then maybe we eventually do it? It's an opportunity to coax some more real-world use cases out of people, and that's a good thing.